keras_spiking/regularizers.py

```diff
@@ -65,137 +139,127 @@ class L2(L1L2):
-           [
-               tf.keras.layers.TimeDistributed(
-                   tf.keras.layers.Dense(50), input_shape=(None, 2)
-               ),
-               keras_spiking.SpikingActivation(
-                   "relu",
-                   dt=1,
-                   spiking_aware_training=False,
-                   activity_regularizer=keras_spiking.PercentileRange(
-                       scale=0.01, minimum=5, maximum=10, penalty="l2"
-                   ),
-               ),
-               tf.keras.layers.GlobalAveragePooling1D(),
-               tf.keras.layers.Dense(1),
-           ]
-       )
-       model.compile(
-           optimizer="rmsprop",
-           loss="mse",
-       )
-       model.fit(train_seq, train_y, epochs=5)
-
-       # create functional model to return spiking layer outputs too
-       func_model = tf.keras.Model(model.input, [model.output, model.layers[1].output])
-       outputs, spikes = func_model.predict(test_seq)
-
-       rates = spikes.mean(axis=1)  # estimate rates by averaging over time
-       max_rates = rates.max(axis=0)
-       print("Max rates: %s, %s" % (max_rates.mean(), max_rates.std()))
-
-       error = np.mean(np.abs(outputs - test_y))
-       print("Error: %s" % (error,))
+        train_x = np.random.uniform(-1, 1, size=(1024 * 100, 2))
+        train_y = train_x[:, :1] * train_x[:, 1:]
+        test_x = np.random.uniform(-1, 1, size=(128, 2))
+        test_y = test_x[:, :1] * test_x[:, 1:]
+
+        # train using one timestep, to speed things up
+        train_seq = train_x[:, None]
+
+        # test using 10 timesteps
+        n_steps = 10
+        test_seq = np.tile(test_x[:, None], (1, n_steps, 1))
+
+        inp = x = tf.keras.Input((None, 2))
+        x = tf.keras.layers.TimeDistributed(tf.keras.layers.Dense(50))(x)
+        x = spikes = keras_spiking.SpikingActivation(
+            "relu",
+            dt=1,
+            activity_regularizer=keras_spiking.regularizers.Percentile(target=(5, 10)),
+        )(x)
+        x = tf.keras.layers.GlobalAveragePooling1D()(x)
+        x = tf.keras.layers.Dense(1)(x)
+
+        model = tf.keras.Model(inp, (x, spikes))
+
+        model.compile(
+            optimizer="rmsprop", loss={"dense_1": "mse"}, metrics={"dense_1": "mae"}
+        )
+        model.fit(train_seq, train_y, epochs=5)
+
+        outputs, spikes = model.predict(test_seq)
+
+        # estimate rates by averaging over time
+        rates = spikes.mean(axis=1)
+        max_rates = rates.max(axis=0)
+        print("Max rates: %s, %s" % (max_rates.mean(), max_rates.std()))
+
+        error = np.mean(np.abs(outputs - test_y))
+        print("MAE: %s" % (error,))
+
 
     .. testoutput::
        :hide:
-       :options: +ELLIPSIS, +NORMALIZE_WHITESPACE
 
        ...
        Max rates: ...
-       Error: ...
-
-    We use ``spiking_aware_training=False``, since activity regularizers perform better
-    on continuous activations (firing rates) than they do on discrete spikes.
```

Comment:
In the case of this example, you found that it worked fine with `spiking_aware_training=True` works fine, but in general it doesn't. I ran into a number of cases where the optimization wasn't converging with it on. Also, the loss will be much higher with it on, because even if the target rate is 500 Hz and the neuron is spiking every second timestep, the output values will be [1000, 0, 1000, 0] which will be penalized.

In general, we really do want to be averaging over time when we apply regularizers, I think, if `spiking_aware_training=True`. Without something to average over time, I think it's best to recommend keeping it off with regularizers. I've found regularizers to be really finicky for changing spike rates (partly to do with the lack of special initialization to get firing rates in the right ballpark), and I'm worried that users would take this example and put it into their own code and find it doesn't work. Even something as simple as changing the amount of regularization slightly can cause it to not converge.

Reply:
I also tried setting `spiking_aware_training=False` in the notebook where we apply L2 regularization, and it made it worse. So I don't think it's clear which option is better in general, which is why I opted to just not bring it up here.

Reply:
In that case, I would still mention the issue here, and say it's unclear which is better and an area of active research (or something like that). I want to make it clear to users that this is not a copy-paste one-size-fits-all example, and for their own networks, they might have to play around.

Maybe this isn't the best place for that, though, and we should add something to the Sphinx docs. We could then point to that here without going into detail.

It would also be good to mention the initialization issue somewhere (i.e. that if you can initialize your network with firing rates that are closer to the target values, optimization can be easier). It *will* make users say "how do you do that?", but I think it's better to have it out in the open, rather than not mention it and then have users frustrated if they try to use our regularizers and they don't work.

Reply:
I added a note on both these points to the regularization section in the example (it seemed like too much text to try to fit into the docstring example).